### PR TITLE
feat(credentials): implement matchExpression-based credentials

### PR DESCRIPTION
**Pull request by andrewazores**
Related to #895
Depends on https://github.com/cryostatio/cryostat-web/pull/475

This PR enhances the existing CredentialsManager by replacing the `targetId` field of stored credentials with a `matchExpression`. This uses the same expression evaluation internal infrastructure already in place for Automated Rules, so the syntax and behaviour is exactly the same. There is a migration function that checks all stored credentials already on disk and checks if they have the `targetId` field. If they do not then they are skipped over, since they are either invalid files or conform to the new format already.

```json
{
  "targetId": "foo",
  "credentials": {
    "username": "user",
    "password": "pass"
  }
}
```

is migrated by applying a simple transformation to:

```json
{
  "matchExpression": "target.connectUrl == \"foo\"",
  "credentials": {
    "username": "user",
    "password": "pass"
  }
}
```

This allows the new expression-based system to migrate existing credentials while maintaining the same semantics, and also allows the old credentials API endpoints to continue working with the same behaviour.

Still TODO in follow-up PRs:
1. Add new API endpoints to allow clients to define and delete credentials with generalized `matchExpression`, not only the backward-compatible `target.connectUrl == "foo"` style
2. ~~Use the new `CredentialsGetHandler` on the frontend to enhance the Security view to better reflect these expression-based credentials~~ https://github.com/cryostatio/cryostat-web/issues/465

Since it is now possible for one set of credentials to match multiple targets, the `AbstractAuthenticatedRequestHandler` no longer deletes credentials if they are used for a JMX connection and the connection fails - since the credentials may apply to multiple targets, they may be valid for other targets than the one that was just checked.

It is also possible that one target may be matched by multiple sets of credentials. There is no protection against this case, so it would simply be considered a user configuration error. If this occurs then the `credentialsManager` will simply provide the "first" matching set of credentials for a given target when requested. The ordering of credentials is intentionally undefined at this time. Multiple credentials for a given target is not an expected valid configuration and so no scheme of trying each of them in order is attempted, or any other more complex behaviours.

### CODE REVIEW

**Review by jan-law — APPROVED** [review]
Works well

**Comment by andrewazores** [comment]
Actually, hold on. I'll add unit tests for the CredentialsManager. It probably should have had them before and it certainly should now.

**Comment by andrewazores** [comment]
Not done adding tests yet, just added basic tests for the new migration function so far. I'll add some more tomorrow.

**Review by hareetd — COMMENTED** [review]
I haven't reviewed the tests yet but the implementation changes look good. However, since the credentials-related notifications have been changed to accommodate the match expression change, the `StoreJmxCredentials` component no longer updates the table state properly when credentials are added or deleted.

**Comment by andrewazores** [comment]
Yea, I'll fix that with TODO 2 listed in the PR body.

**Comment by andrewazores** [comment]
Filing the `-web` side to do later, since I'm on PTO next week and won't have any time to get started on it. I'll leave it open for anyone else to take a stab at if there's interest, otherwise I'll start on it when I'm back to work.

File: src/test/java/io/cryostat/configuration/CredentialsManagerTest.java
Comment (hareetd) [inline src/test/java/io/cryostat/configuration/CredentialsManagerTest.java:452]: Could you explain the difference between these two tests? I noticed that the second one doesn't require `Mockito.when(platformClient.listDiscoverableServices()).thenReturn(List.of(serviceRef));` even though they seem functionally identical.

Comment (andrewazores) [reply src/test/java/io/cryostat/configuration/CredentialsManagerTest.java:452]: Between the two `Loading` tests? One calls `credentialsManager.getCredentials(ServiceRef serviceRef)` and the other calls `credentialsManager.getCredentialsByTargetId(String targetId)`. The `byTargetId` one is there to preserve backward-compatibility with the older API and the `AbstractAuthenticatedRequestHandler`, so that the new style can be used for API paths like `/api/v1/targets/:targetId/recordings`, where we don't have a full `ServiceRef` object to evaluate the `matchExpression` against. In that case, the credentials manager searches for known targets that match the given `targetId`, maps that to a `ServiceRef` if it can, and then checks for matching credentials for that. The `getCredentials(ServiceRef)` just directly looks for credentials that match the supplied `ServiceRef`, without having to call out to the platform client to query for targets that match (since this is already a reference to a target).

Comment (hareetd) [reply src/test/java/io/cryostat/configuration/CredentialsManagerTest.java:452]: Makes sense, thanks. I should've just diffed the two files instead of looking to see what they differences were; missed the `getCredentials(...)` vs. `getCredentialsByTargetId(...)` calls :P

**Review by hareetd — APPROVED** [review]

**Comment by jan-law** [comment]
Backend looks good, I'm assuming you'll merge this once the web client gets updated

**Comment by andrewazores** [comment]
`-web` side merged and updated in this feature branch. Nothing else has changed since previous review approvals.

**Review by hareetd — APPROVED** [review]

**Review by hareetd — COMMENTED** [review]
Looks like the web-client submodule hasn't been updated with the frontend changes.

**Comment by andrewazores** [comment]
It isn't listed as a commit in this PR since I rebased on top, but the latest frontend change in the submodule is already in `main` thanks to our new CI-bot: https://github.com/cryostatio/cryostat/commit/d4b448ccc598cefc6d02e186b45b82aae5360db6

**Comment by hareetd** [comment]
Oh that makes sense then, cool.

**Comment by andrewazores** [comment]
No more manually updating the submodule :-)

**Pull request merged**
